crates/bounded_planet_core/src/camera/mod.rs

```diff
@@ -197,64 +197,108 @@ impl UniversalGeometry {
-    /// original position `o`, original rotation `r`, and scroll weight `s`.
-    fn zoom(&self, p: Vec3, r: Quat, s: f32, scale: f32) -> (Vec3, Quat) {
-        fn plane(_o: Vec3, n: Vec3, p: Vec3, r: Quat, s: f32, scale: f32) -> (Vec3, Quat) {
-            let mut delta = (-r).mul_vec3(Vec3::new(0.0, 0.0, s)); // unscaled delta
-            delta *= scale * p.dot(n).abs(); // scale delta by dist
+    /// Get the new transformation resulting from the original [`Transform`]
+    /// `ot` and scroll weight `s`.
+    fn noparent_zoom(&self, ot: Transform, s: f32, scale: f32) -> Transform {
+        fn plane(_o: Vec3, n: Vec3, mut ot: Transform, s: f32, scale: f32) -> Transform {
+            let mut delta = (-ot.rotation()).mul_vec3(Vec3::new(0.0, 0.0, s)); // unscaled delta
+            delta *= scale * ot.translation().dot(n).abs(); // scale delta by dist
 
-            (delta, Quat::identity())
+            let oti = ot.value_mut();
+            *oti = Mat4::from_translation(delta) * *oti;
+            ot
         }
 
-        match self {
-            UniversalGeometry::Plane { origin, normal } => plane(origin.0, *normal, p, r, s, scale),
+        match self.0 {
+            UniversalGeometry::Plane { origin, normal } => plane(origin, normal, ot, s, scale),
         }
     }
-}
 
-/// A private newtype of Universal Geometry, that satisfies some invariants:
-/// 1) `self.0` is normalized.
-/// 2) If `self.0` is a plane, then its normal is a unit vector.
-#[derive(Copy, Clone)]
-struct InternalUG(UniversalGeometry);
+    /// Return the `(parent, camera)` new transformations resulting from the
+    /// original [`Transform`]s `opt` and `oct` of the parent and camera
+    /// respectively and movement `s` in terms of [`InternalUG`] space.
+    fn parent_trans(
+        &self,
+        opt: Transform,
+        oct: Transform,
+        s: Vec3,
+        scale: f32
+    ) -> (Transform, Transform) {
+        match self.0 {
+            UniversalGeometry::Plane {..} => {
+                let mut out = self.noparent_trans(Transform::new(*oct.value()), s, scale);
+                let outi = out.value_mut();
+                // insert appropriate logic here
+                (out, oct)
+            }
+        }
+    }
 
-impl From<UniversalGeometry> for InternalUG {
-    fn from(ug: UniversalGeometry) -> Self {
-        InternalUG(ug)
+    /// Return the `(parent, camera)` new transformations resulting from the
+    /// original [`Transform`]s `opt` and `oct` of the parent and camera
+    /// respectively and and scroll weight `s`.
+    fn parent_zoom(
+        &self,
+        opt: Transform,
+        oct: Transform,
+        s: f32,
+        scale: f32
+    ) -> (Transform, Transform) {
+        match self.0 {
+            UniversalGeometry::Plane {..} => {
+                let mut out = self.noparent_zoom(Transform::new(*oct.value()), s, scale);
+                let outi = out.value_mut();
+                // insert appropriate logic here
```

Comment:
use `//todo(#issue_number)` here

Reply:
See the description, this is the logic I haven't finished.